core/coreunix/reed_solomon_add.go

```diff
@@ -0,0 +1,211 @@
+	dag "github.com/ipfs/go-merkledag"
+	gopath "path"
+)
+
+type ReedSolomonAdder struct {
+	Adder
+	InfileReaders []io.Reader
+	IsDir         bool
+	CurrentOffset uint64
+}
+
+// NewReedSolomonAdder returns a new ReedSolomonAdder used for a file/directory add operation.
+func NewReedSolomonAdder(adder *Adder) (*ReedSolomonAdder, error) {
+	rsa := &ReedSolomonAdder{
+		InfileReaders: nil,
+		IsDir:         false,
+		CurrentOffset: 0,
+	}
+	rsa.Adder = *adder
+
+	return rsa, nil
+}
+
+// AddAllAndPin adds the given request's files and pin them.
+func (rsadder *ReedSolomonAdder) AddAllAndPin(file files.Node) (ipld.Node, error) {
+	if rsadder.Pin {
+		rsadder.unlocker = rsadder.gcLocker.PinLock()
+	}
+	defer func() {
+		if rsadder.unlocker != nil {
+			rsadder.unlocker.Unlock()
+		}
+	}()
+
+	// Create a directory tree from the input `file` directory.
+	n, err := rsadder.addFileNode("", file, true)
+	if err != nil {
+		return nil, err
+	}
+	var byts []byte
+	switch node := n.(type) {
+	case *uio.DirNode, *uio.SymlinkNode, *uio.FileNode:
+		byts, err = json.Marshal(node)
+		if err != nil {
+			return nil, err
+		}
+	default:
+		return nil, errors.New("unexpected node type")
+	}
+
+	var reader io.Reader = io.MultiReader(rsadder.InfileReaders...)
+	if rsadder.Progress {
+		reader = &progressReader{file: reader, path: "", out: rsadder.Out}
+	}
+
+	// Create a DAG with the above directory tree as metadata and
+	// the data from the given `file` directory or file.
+	nd, err := rsadder.add(reader, byts)
+	if err != nil {
+		return nil, err
+	}
+
+	// output directory and file events
+	err = rsadder.outputDirs("", nd)
+	if err != nil {
+		return nil, err
+	}
+
+	// Pin the newly created DAG.
+	if !rsadder.Pin {
+		return nd, nil
+	}
+
+	if err := rsadder.PinRoot(nd); err != nil {
+		return nil, err
+	}
+
+	return nd, nil
+}
+
+// addFileNode traverses the directory tree under
+// the given `file` in a bottom up DFS way.
+func (rsadder *ReedSolomonAdder) addFileNode(path string, file files.Node, toplevel bool) (uio.Node, error) {
+	//defer file.Close()
+
+	err := rsadder.maybePauseForGC()
+	if err != nil {
+		return nil, err
+	}
+
+	if rsadder.liveNodes >= liveCacheSize {
+		// TODO: flush free memory rsadder.mfsRoot()'s FlushMemFree() to flush free memory
+		log.Info("rsadder liveNodes >= liveCacheSize, needs flushing")
+	}
+	rsadder.liveNodes++
+
+	switch f := file.(type) {
+	case files.Directory:
+		return rsadder.addDir(path, f, toplevel)
+	case *files.Symlink:
+		return rsadder.addSymlink(path, f)
+	case files.File:
+		return rsadder.addFile(path, f)
+	default:
+		return nil, errors.New("unknown file type")
+	}
+}
+
+func (rsadder *ReedSolomonAdder) addDir(path string, dir files.Directory, toplevel bool) (uio.Node, error) {
+	log.Infof("adding directory: %s", path)
+
+	_, dstName := gopath.Split(path)
+	node := &uio.DirNode{
+		BaseNode: uio.BaseNode{
+			NodeType: uio.DirNodeType,
+			NodePath: path,
+			NodeName: dstName},
+		//Directory: dir,
```

Comment:
why is it not setting dir here? can you briefly mention when (if somewhere) this will be set?

Reply:
Added comments.